docs/en/query_language/functions/json_functions.md

```diff
@@ -55,5 +55,112 @@ visitParamExtractString('{"abc":"hello}', 'abc') = ''
 
 There is currently no support for code points in the format `\uXXXX\uYYYY` that are not from the basic multilingual plane (they are converted to CESU-8 instead of UTF-8).
 
+# New functions for working with JSON
+
+## Motivation
+Storing JSON in a String column is fudamentally against the columnstore nature of Clickhouse 
+and is almost always a wrong choice leading to performance problems. On the other hand, the maximum level 
+of nested structures is limited to 2, and modelling 1 to M relationships using several relational tables
+is also not always possible in an efficient way.
+
+Another use case for JSONs in String columns could be a generic storage system receiving
+JSON messages in unspecified and persisting them; further data processing could be done by
+parsing the JSON using DEFAULT columns and MATERIALIZED VIEWS.
+
+The new JSON functions aim to solve some limitations of the existing JSON functions, 
+without pretending to provide a full JSON support or trying to make Clickhouse into a 
+document store. Usage of JSON in Strings must remain an exception.
+
+The following limitations have been solved:
+- Cannot extract JSON arrays
+- Cannot provide a path of property names for extraction
+- Cannot extract all matches of the key (only the first one is returned)
+- Function names are long and hard to remember
```

Comment:
I think `visitParamExtractXXX` is probably better, it has more explicit semantics.

Reply:
I agree it is better in the situations similar to Yandex Metrika, where the JSON has a relatively simple structure and comes from the outside world and therefore could be malformed or maliciously constructed to exploit some security vulnerability. 
In the situations like ours JSON is generated in-house, and has a pretty complicated structure, so you would probably need to call several visitParamsExtractRaw one after another to get to the data, which is quite long to read.

Reply:
Yes, it's a great job, I just thought it might be better to rename `jsonAny ` to `visitParamExtractAny` ?

Reply:
Well, the old visitParamExtract functions are still there and can be used. I'm not quite sure that I understand what negative consequence of additional json functions do you see?

Reply:
Shorter function names was an important goal why we've started working on this; so we can't just rename it. Function names like visitParamExtractAllArrays are to long, and we would need to use several calls of them per query. How can we move forward with the PR?
- Implement plugin architecture so we can drop the .so file with our functions after installing the standard Clickhouse version?
- Implement a way to create aliases of function names in server configuration file?
- Merge our functions into Clickhouse master, but leave them undocumented, so only small fraction of users will use them and depend on them?